File: 377_CombinationSumIV.py
# In order to find all possible combinations that equal target, memoized backtracking solution
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        self.cache = {}
        
        def helper(arr, curr):
            if curr > target:
                return 0
            elif curr == target:
                return 1
            elif curr in self.cache:
                return self.cache[curr]
            
            total = 0
            for i in range(len(arr)):
                if curr+arr[i] not in self.cache:
                    self.cache[curr+arr[i]] = helper(arr, curr+arr[i])
                total += self.cache[curr+arr[i]]
            
            return total
        
        return helper(nums, 0)

File: cpp/377_CombinationSumIV.cpp
class Solution {
public:
    unordered_map<int, int> cache;
    int combinationSum4(vector<int>& nums, int target) {
        return combine(0, nums, target);
    }
    
    int combine(int curr_sum, vector<int>& nums, int target) {
        if (curr_sum == target)
            return 1;
        else if (curr_sum > target)
            return 0;
        else if (cache.find(curr_sum) != cache.end())
            return cache[curr_sum];
            
        int total = 0;
        for (int i = 0; i < nums.size(); ++i) {
            int add_element = curr_sum + nums[i];
            if (cache.find(add_element) == cache.end())
                cache[add_element] = combine(add_element, nums, target);
            total += cache[add_element];
        }
        return total;
    }
};
